I18n reg forms

danielglauser opened the pull request:
[Card here](https://www.pivotaltracker.com/story/show/117971281), adds support for a Spanish NVRF form.

cap10morgan commented on resources/config.edn:
```diff
@@ -15,4 +15,5 @@
             :election-authority-works-search 10000}
- :templates {:nvrf "https://s3-us-west-2.amazonaws.com/voter-registration-works/nvrf.en.pdf"}}
+  :templates {:nvrf {:en "https://s3-us-west-2.amazonaws.com/voter-registration-works/nvrf.en.pdf"
+                     :es "https://s3-us-west-2.amazonaws.com/voter-registration-works/nvrf.es.pdf"}}}
```
Can we scoot the indentation back one?

cap10morgan replied:
Still need to fix this.

cap10morgan commented on src/voter_registration_works/handlers.clj:
```diff
@@ -109,7 +109,7 @@
            (s/optional-key :country-of-birth) s/Str ; TODO: Replace with OCDCountry once that's in here
            (s/optional-key :race) s/Str ; TODO: Replace with TVEthnicity once that's in here
-           (s/optional-key :language-preference) s/Str ; TODO: Replace with TVLocale once that's in here
+           (s/optional-key :language-preference) (s/enum :en :es)
```
I'm not sure we want to constrain this here. California's OVR API, for example, accepts more values than this, and we may want to pass those along before our site has translations for them. Also this should be a generic API that's not tied too tightly to turbovote-web, so it shouldn't know/care that it only supports English and Spanish.

Also, this was a string before, according to the schema. Should we be coercing the `:en` and `:es` values into something else?

cap10morgan commented on src/voter_registration_works/reg_form.clj:
```diff
@@ -71,6 +71,44 @@
+          :race (:race voter)
+          :phone (:phone voter)
+          :dob (:date-of-birth voter)}
+         (condp = (:language-preference voter)
```
Can you add a comment about why this language branch logic is necessary here? I think I would find it very surprising if I didn't know about the fun NVRF form differences.

cap10morgan commented on src/voter_registration_works/reg_form.clj:
```diff
@@ -131,7 +148,7 @@
   (wire-up/async->fn channels/render-form))
 
-(defn generate-registration-form [registration]
+(defn generate-registration-form [registration] ;TODO need to make sure language gets here...
```
Seems like we should do this now?

cap10morgan commented on test/voter_registration_works/reg_form_test.clj:
```diff
@@ -7,26 +7,45 @@
+
+(def nvrf-es-form-message-required-fields
+  {:template (config [:templates :nvrf :es])
+   :name "Ramón_Gonzales_nvrf.pdf"
```
Should the accent be in the filename here? If the form-works consumer is going to tell it what filename to use, shouldn't form-works just do it (versus having form-works translate non-ASCII chars to their nearest ASCII equivalents)?

I'm not 100% clear on why we can't use non-ASCII chars in filenames, but I trust that the reasons are real. I just wonder if it should be form-works job to enforce that or if it should be the responsibility of its consumers.

Doesn't have to hold this up, just a thought that occurred to me.

danielglauser replied:
@cap10morgan The reason non-accent characters are bad is when the file is used as an email attachment. That's the purpose of the code that cleans the name featured in [this](https://github.com/democracyworks/voter-registration-works/pull/26) PR. We'll have to be careful when merging this into master to make sure it doesn't overwrite the existing functionality. Luckily the tests will break if we screw that up.

As far as `form-works` enforcing the ASCII filenames, since the issue is when they are attached to emails it didn't seem like that was the right place to add the code, since `form-works` knows nothing about how the files are used, however it would be nice if we didn't duplicate this functionality across multiple repos.

cap10morgan replied:
I agree that it shouldn't be in form-works and am glad to hear that that's already the case. If it's just an issue with email attachments, should it go into email-works?

danielglauser replied:
@cap10morgan The way `email-works` currently attaches things is by passing a URI to MailJet. For `email-works` to take care of this, it will need to rename the file up on S3.

cap10morgan replied:
No, email-works is just passing all that directly to the postal library. According to [its docs](https://github.com/drewr/postal#attachments) you can set the `:file-name` of the attachment at attachment-time. We shouldn't need to change anything on S3 to do that.

cap10morgan replied:
There may still be reasons to use the same ASCII-ified filenames everywhere, and as you both pointed out, this isn't the PR to discuss it in anyway.

danielglauser commented:
@cap10morgan Done here, please have another look.

cap10morgan commented on test/voter_registration_works/states_test.clj:
```diff
@@ -60,4 +60,7 @@
                (:registration-authority
-                (voter-register :co test-voter :by-mail))))))))
+                (voter-register {:state :co
+                                 :voter test-voter
```
Should we add a `:language-preference` to `test-voter`? Or add a test where one is in there?

danielglauser commented:
@cap10morgan Let me know if that addresses your concerns. We talked about verifying that the correct value for `:language` was passed to the Election Authority lookup, however that would involve putting an `(is ...)` in the `with-redef`'ed function, which seems weird.

cap10morgan commented:
LGTM

danielglauser commented:
@cap10morgan please have a look.
